base/boot.jl

```diff
@@ -496,11 +496,7 @@ const undef = UndefInitializer()
 
 # type and dimensionality specified
 (self::Type{GenericMemory{kind,T,addrspace}})(::UndefInitializer, m::Int) where {T,addrspace,kind} =
-    if isdefined(self, :instance) && m === 0
```

Comment:
This needs some supporting performance analysis, since it hardly even appears in the IR after optimization
```
julia> code_typed(Memory{Int}, (Core.UndefInitializer, Int))
1-element Vector{Any}:
 CodeInfo(
1 ─      nothing::Nothing
│   @ boot.jl:498 within `GenericMemory`
│   %2 = (m === 0)::Bool
└──      goto #3 if not %2
    @ boot.jl:500 within `GenericMemory`
2 ─ %4 = Core.getproperty(self, :instance)::Memory{Int64}
└──      return %4
    @ boot.jl:502 within `GenericMemory`
3 ─ %6 = $(Expr(:foreigncall, :(:jl_alloc_genericmemory), Ref{GenericMemory{kind, T, addrspace}}, svec(Any, Int64), 0, :(:ccall), Core.Argument(1), Core.Argument(3), Core.Argument(3)))::Memory{Int64}
└──      return %6
) => Memory{Int64}
```

Reply:
The entire thing is in the IR.